serializers/src/main/java/io/pravega/schemaregistry/serializers/AbstractDeserializer.java

```diff
@@ -0,0 +1,116 @@
+                                   EncodingCache encodingCache,
+                                   boolean encodeHeader) {
+        Preconditions.checkNotNull(groupId);
+        Preconditions.checkNotNull(client);
+        Preconditions.checkNotNull(encodingCache);
+        this.groupId = groupId;
+        this.client = client;
+        this.encodingCache = encodingCache;
+        this.schemaInfo = schema == null ? null : schema.getSchemaInfo();
+        this.encodeHeader = encodeHeader;
+        this.skipHeaders = skipHeaders;
+        this.decoder = decoder;
+            
+        initialize();
+    }
+
+    private void initialize() {
+        if (schemaInfo != null) {
+            log.info("Validate caller supplied schema.");
+            if (!client.canReadUsing(groupId, schemaInfo)) {
+                throw new IllegalArgumentException("Cannot read using schema" + schemaInfo.getType());
+            }
+        } else {
+            if (!this.encodeHeader) {
+                log.warn("No reader schema is supplied and stream does not have encoding headers.");
+            }
+        }
+    }
+    
+    @SneakyThrows(IOException.class)
+    @Override
+    public T deserialize(ByteBuffer data) {
+        int start = data.arrayOffset() + data.position();
+        if (this.encodeHeader) {
+            SchemaInfo writerSchema = null;
+            ByteBuffer decoded;
+            if (skipHeaders) {
+                data.position(start + HEADER_SIZE);
+                decoded = data;
+            } else {
+                byte protocol = data.get();
+                EncodingId encodingId = new EncodingId(data.getInt());
+                EncodingInfo encodingInfo = encodingCache.getGroupEncodingInfo(encodingId);
+                writerSchema = encodingInfo.getSchemaInfo();
+                decoded = decoder.decode(encodingInfo.getCodecType(), data);
+            }
+
+            ByteArrayInputStream bais = new ByteArrayInputStream(decoded.array(), 
+                    decoded.arrayOffset() + decoded.position(), decoded.remaining());
+            if (schemaInfo == null) { // deserialize into writer schema
+                // pass writer schema for schema to be read into
+                return deserialize(bais, writerSchema, writerSchema);
+            } else {
+                // pass reader schema for schema on read to the underlying implementation
+                return deserialize(bais, writerSchema, schemaInfo);
+            }
+        } else {
+            // pass reader schema for schema on read to the underlying implementation
+            ByteArrayInputStream inputStream = new ByteArrayInputStream(data.array(), 
+                    data.arrayOffset() + data.position(), data.remaining());
+
+            return deserialize(inputStream, null, schemaInfo);
+        }
+    }
+    
+    protected abstract T deserialize(InputStream inputStream, SchemaInfo writerSchema, SchemaInfo readerSchema) throws IOException;
```

Comment:
Both serializers and deserializers were made `Input and Output streams` respectively consciously as per a requirement from claudio. IIUC his intention is to eventually have Pravega's serializer interface also allow for input and output streams because with byte buffers you need to know the size of data before serializing at least.  

And for deserializers, they typically do a one pass on the data and may not require to go back and forth on the data. So inputstream access pattern on the underlying buffer may be ok.